### PR TITLE
Add support for swarm seccomp and apparmor

### PR DESCRIPTION
**Pull request by dperny**
**- What I did**

Added support in Swarm for custom Seccomp profiles, some small amount of AppArmor configuration, and `no-new-privileges`.

**- How I did it**

Swarmkit has the necessary plumbing for these fields added, and now it's wired up in the engine.

**- How to verify it**

**- Description for the changelog**

Added Swarm support for seccomp, AppArmor, and no-new-privileges.

### CODE REVIEW

File: daemon/cluster/executor/container/container.go [hunk 1]
```diff
@@ -718,6 +718,36 @@ func (c *containerConfig) applyPrivileges(hc *enginecontainer.HostConfig) {
 			hc.SecurityOpt = append(hc.SecurityOpt, "label=type:"+selinux.Type)
 		}
 	}
+
+	// variable to make the lines shorter and easier to read
+	seccomp := privileges.Seccomp
+	if seccomp != nil {
+		switch seccomp.Mode {
+		case api.Privileges_SeccompOpts_DEFAULT:
+			// TODO(dperny): DO NOT MERGE WITHOUT THIS QUESTION ANSWERED!
```
Comment (corhere) [inline daemon/cluster/executor/container/container.go:727]: Should this PR be converted to draft until this question is answered?

Comment (dperny) [reply daemon/cluster/executor/container/container.go:727]: This review comment is why I put this code comment. Knew I would forget about it.

File: api/types/swarm/container.go [hunk 2]
```diff
@@ -41,8 +77,11 @@ type CredentialSpec struct {
 
 // Privileges defines the security options for the container.
 type Privileges struct {
-	CredentialSpec *CredentialSpec
-	SELinuxContext *SELinuxContext
+	CredentialSpec  *CredentialSpec
+	SELinuxContext  *SELinuxContext
+	Seccomp         *SeccompOpts
+	AppArmor        *AppArmorOpts
+	NoNewPrivileges bool
```
Comment (thaJeztah) [inline api/types/swarm/container.go:84]: Do we want these to have an `omitempty` ?  (also for older API versions?)

File: daemon/cluster/executor/container/container.go [hunk 3]
```diff
@@ -718,6 +718,34 @@ func (c *containerConfig) applyPrivileges(hc *enginecontainer.HostConfig) {
 			hc.SecurityOpt = append(hc.SecurityOpt, "label=type:"+selinux.Type)
 		}
 	}
+
+	// variable to make the lines shorter and easier to read
+	seccomp := privileges.Seccomp
+	if seccomp != nil {
```
Comment (thaJeztah) [inline daemon/cluster/executor/container/container.go:724]: Minor nit; might as well keep it fully local to this `if`;

```suggestion
	if seccomp := privileges.Seccomp; seccomp != nil {
```

File: daemon/cluster/executor/container/container.go [hunk 4]
```diff
@@ -718,6 +718,34 @@ func (c *containerConfig) applyPrivileges(hc *enginecontainer.HostConfig) {
...
+			// basically verbatim what happens in the cli after a file is read.
+			hc.SecurityOpt = append(hc.SecurityOpt, fmt.Sprintf("seccomp=%s", seccomp.Profile))
+		}
+	}
+
+	apparmor := privileges.Apparmor
+	// if the setting is DEFAULT, then nothing to be done. If it's DISABLED,
+	// we set that. Custom not supported yet. When custom *is* supported, make
+	// it look like the above.
+	if apparmor != nil && apparmor.Mode == api.Privileges_AppArmorOpts_DISABLED {
```
Comment (thaJeztah) [inline daemon/cluster/executor/container/container.go:742]: Same here;

```suggestion
	// if the setting is DEFAULT, then nothing to be done. If it's DISABLED,
	// we set that. Custom not supported yet. When custom *is* supported, make
	// it look like the above.
	if apparmor := privileges.Apparmor; apparmor != nil && apparmor.Mode == api.Privileges_AppArmorOpts_DISABLED {
```

**Review by thaJeztah — COMMENTED** [review]
some very minor comments, also could you add entries to https://github.com/moby/moby/blob/master/docs/api/version-history.md to mention the new options (which apply to both `POST /services/create` and `POST /services/{id}/update`)

File: daemon/cluster/convert/container.go [hunk 5]
```diff
@@ -69,6 +69,34 @@ func containerSpecFromGRPC(c *swarmapi.ContainerSpec) *types.ContainerSpec {
 				Level:   c.Privileges.SELinuxContext.Level,
 			}
 		}
+
+		if c.Privileges.Seccomp != nil {
```
Comment (thaJeztah) [inline daemon/cluster/convert/container.go:73]: I guess ideally we'd have some tests added as well, but I'm ok with doing it in a follow-up if we'd like to have this included in the release. I see there's some tests in this package already; https://github.com/moby/moby/tree/78479b19158290e7338ceb3985bb809b83de5fd4/daemon/cluster/executor/container

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
ah, dang; needs a rebase, @dperny (`version-history.md`)

**Review by neersighted — APPROVED** [review]

**Comment by Qwarctick** [comment]
Hello @dperny.
Thanks for your PR.
How does this feature is suppose to work ? With the CLI ? With the API ? With the compose file ?

Thanks in advance :)

**Comment by neersighted** [comment]
This is purely backend support; the issue you already linked is the appropriate venue to discuss UI.

**Pull request merged**